File: audit.py
#!/usr/bin/env python3
"""
Author: KleaSCM
Email: KleaSCM@gmail.com
Name: audit_assets.py
Description: Comprehensive asset library auditing script that counts assets, lists directories, and provides detailed statistics
"""

import os
import sys
from pathlib import Path
from collections import defaultdict
import json
from datetime import datetime

class AssetAuditor:
    def __init__(self):
        self.project_root = Path(__file__).parent
        self.assets_path = self.project_root / "Assets"
        self.stats = {
            'total_files': 0,
            'total_directories': 0,
            'file_types': defaultdict(int),
            'directories': [],
            'largest_files': [],
            'missing_files': [],
            'duplicate_names': [],
            'categories': defaultdict(int),
            'asset_types': {
                'models': 0,
                'textures': 0,
                'audio': 0,
                'video': 0,
                'documents': 0,
                'archives': 0,
                'scripts': 0,
                'other': 0
            },
            'size_breakdown': {
                'tiny': 0,      # < 1 MB
                'small': 0,     # 1-10 MB
                'medium': 0,    # 10-100 MB
                'large': 0,     # 100 MB - 1 GB
                'huge': 0,      # 1-10 GB
                'massive': 0    # > 10 GB
            }
        }
        
    def get_file_size_mb(self, file_path):
        """Get file size in MB - handles ANY size"""
        try:
            return os.path.getsize(file_path) / (1024 * 1024)
        except:
            return 0
    
    def get_file_size_gb(self, file_path):
        """Get file size in GB - for massive files"""
        try:
            return os.path.getsize(file_path) / (1024 * 1024 * 1024)
        except:
            return 0
    
    def format_file_size(self, size_mb):
        """Format file size for display - handles ANY size"""
        if size_mb >= 1024 * 1024:  # > 1 TB
            return f"{size_mb / (1024 * 1024):.2f} TB"
        elif size_mb >= 1024:  # > 1 GB
            return f"{size_mb / 1024:.2f} GB"
        elif size_mb >= 1:  # > 1 MB
            return f"{size_mb:.2f} MB"
        else:  # < 1 MB
            return f"{size_mb * 1024:.0f} KB"
    
    def categorize_file_by_type(self, file_path, ext, size_mb):
        """Categorize file by type and size"""
        # Universal file type detection
        model_extensions = {'.blend', '.obj', '.fbx', '.dae', '.3ds', '.stl', '.ply', '.max', '.c4d', '.ma', '.mb', '.abc', '.usd', '.gltf', '.glb'}
        texture_extensions = {'.png', '.jpg', '.jpeg', '.tga', '.tiff', '.bmp', '.exr', '.hdr', '.psd', '.ai', '.svg', '.webp', '.ktx', '.dds'}
        audio_extensions = {'.mp3', '.wav', '.flac', '.aac', '.ogg', '.wma', '.m4a', '.aiff', '.au', '.mid', '.midi'}
        video_extensions = {'.mp4', '.avi', '.mov', '.wmv', '.flv', '.webm', '.mkv', '.m4v', '.3gp', '.ogv', '.ts', '.mts'}
        document_extensions = {'.pdf', '.doc', '.docx', '.txt', '.rtf', '.md', '.html', '.xml', '.json', '.csv', '.xlsx', '.ppt', '.pptx'}
        archive_extensions = {'.zip', '.rar', '.7z', '.tar', '.gz', '.bz2', '.xz', '.dmg', '.iso'}
        script_extensions = {'.py', '.js', '.php', '.rb', '.java', '.cpp', '.c', '.cs', '.sh', '.bat', '.ps1'}
        
        # Categorize by type
        if ext in model_extensions:
            self.stats['asset_types']['models'] += 1
        elif ext in texture_extensions:
            self.stats['asset_types']['textures'] += 1
        elif ext in audio_extensions:
            self.stats['asset_types']['audio'] += 1
        elif ext in video_extensions:
            self.stats['asset_types']['video'] += 1
        elif ext in document_extensions:
            self.stats['asset_types']['documents'] += 1
        elif ext in archive_extensions:
            self.stats['asset_types']['archives'] += 1
        elif ext in script_extensions:
            self.stats['asset_types']['scripts'] += 1
        else:
            self.stats['asset_types']['other'] += 1
        
        # Categorize by size
        if size_mb < 1:
            self.stats['size_breakdown']['tiny'] += 1
        elif size_mb < 10:
            self.stats['size_breakdown']['small'] += 1
        elif size_mb < 100:
            self.stats['size_breakdown']['medium'] += 1
        elif size_mb < 1024:  # 1 GB
            self.stats['size_breakdown']['large'] += 1
        elif size_mb < 10240:  # 10 GB
            self.stats['size_breakdown']['huge'] += 1
        else:
            self.stats['size_breakdown']['massive'] += 1
    
    def check_for_missing_references(self, file_path, ext):
        """Universal missing file reference detection"""
        # Common reference patterns
        reference_patterns = {
            '.mtl': ['.obj'],  # Material files reference OBJ files
            '.blend': ['.png', '.jpg', '.jpeg', '.tga', '.tiff'],  # Blender files reference textures
            '.max': ['.png', '.jpg', '.jpeg', '.tga', '.tiff'],  # 3DS Max files reference textures
            '.ma': ['.png', '.jpg', '.jpeg', '.tga', '.tiff'],  # Maya files reference textures
            '.mb': ['.png', '.jpg', '.jpeg', '.tga', '.tiff'],  # Maya files reference textures
            '.c4d': ['.png', '.jpg', '.jpeg', '.tga', '.tiff'],  # Cinema 4D files reference textures
            '.fbx': ['.png', '.jpg', '.jpeg', '.tga', '.tiff'],  # FBX files might reference textures
            '.usd': ['.png', '.jpg', '.jpeg', '.tga', '.tiff'],  # USD files might reference textures
            '.gltf': ['.png', '.jpg', '.jpeg', '.tga', '.tiff'],  # glTF files reference textures
            '.glb': ['.png', '.jpg', '.jpeg', '.tga', '.tiff'],  # glTF binary files reference textures
        }
        
        if ext in reference_patterns:
            for ref_ext in reference_patterns[ext]:
                # Check if referenced files exist in same directory
                ref_file = file_path.with_suffix(ref_ext)
                if not ref_file.exists():
                    # Also check in subdirectories
                    found = False
                    for subdir in file_path.parent.iterdir():
                        if subdir.is_dir():
                            potential_ref = subdir / (file_path.stem + ref_ext)
                            if potential_ref.exists():
                                found = True
                                break
                    
                    if not found:
                        self.stats['missing_files'].append(str(ref_file.relative_to(self.project_root)))
    
    def scan_directory(self, directory):
        """Recursively scan directory and collect statistics - UNIVERSAL VERSION"""
        if not directory.exists():
            print(f"❌ Directory not found: {directory}")
            return
            
        print(f"🔍 Scanning: {directory}")
        
        # Universal file type categories
        model_extensions = {'.blend', '.obj', '.fbx', '.dae', '.3ds', '.stl', '.ply', '.max', '.c4d', '.ma', '.mb', '.abc', '.usd', '.gltf', '.glb'}
        texture_extensions = {'.png', '.jpg', '.jpeg', '.tga', '.tiff', '.bmp', '.exr', '.hdr', '.psd', '.ai', '.svg', '.webp', '.ktx', '.dds'}
        audio_extensions = {'.mp3', '.wav', '.flac', '.aac', '.ogg', '.wma', '.m4a', '.aiff', '.au', '.mid', '.midi'}
        video_extensions = {'.mp4', '.avi', '.mov', '.wmv', '.flv', '.webm', '.mkv', '.m4v', '.3gp', '.ogv', '.ts', '.mts'}
        document_extensions = {'.pdf', '.doc', '.docx', '.txt', '.rtf', '.md', '.html', '.xml', '.json', '.csv', '.xlsx', '.ppt', '.pptx'}
        archive_extensions = {'.zip', '.rar', '.7z', '.tar', '.gz', '.bz2', '.xz', '.dmg', '.iso'}
        script_extensions = {'.py', '.js', '.php', '.rb', '.java', '.cpp', '.c', '.cs', '.sh', '.bat', '.ps1'}
        
        for root, dirs, files in os.walk(directory):
            root_path = Path(root)
            
            # Count directories
            self.stats['total_directories'] += len(dirs)
            
            # Add directory to list
            for dir_name in dirs:
                full_dir_path = root_path / dir_name
                self.stats['directories'].append(str(full_dir_path.relative_to(self.project_root)))
                
                # UNIVERSAL categorization - detect ANY category automatically
                dir_lower = dir_name.lower()
                
                # Auto-detect categories from directory names
                if any(keyword in dir_lower for keyword in ['model', 'building', 'character', 'prop', 'mesh', 'geometry']):
                    self.stats['categories']['Models'] += 1
                elif any(keyword in dir_lower for keyword in ['texture', 'material', 'map', 'diffuse', 'normal', 'specular']):
                    self.stats['categories']['Textures'] += 1
                elif any(keyword in dir_lower for keyword in ['audio', 'sound', 'music', 'voice', 'sfx']):
                    self.stats['categories']['Audio'] += 1
                elif any(keyword in dir_lower for keyword in ['video', 'movie', 'animation', 'render', 'footage']):
                    self.stats['categories']['Video'] += 1
                elif any(keyword in dir_lower for keyword in ['scene', 'level', 'environment', 'world']):
                    self.stats['categories']['Scenes'] += 1
                elif any(keyword in dir_lower for keyword in ['script', 'code', 'programming']):
                    self.stats['categories']['Scripts'] += 1
                elif any(keyword in dir_lower for keyword in ['doc', 'manual', 'guide', 'readme']):
                    self.stats['categories']['Documents'] += 1
                elif any(keyword in dir_lower for keyword in ['archive', 'backup', 'compressed']):
                    self.stats['categories']['Archives'] += 1
                else:
                    # Auto-detect from directory name
                    self.stats['categories'][dir_name.title()] += 1
            
            # Process files
            for file in files:
                file_path = root_path / file
                self.stats['total_files'] += 1
                
                # Get file extension
                ext = file_path.suffix.lower()
                self.stats['file_types'][ext] += 1
                
                # Get file size (handles ANY size)
                size_mb = self.get_file_size_mb(file_path)
                
                # Track largest files
                self.stats['largest_files'].append({
                    'path': str(file_path.relative_to(self.project_root)),
                    'size_mb': size_mb
                })
                
                # UNIVERSAL missing file detection
                self.check_for_missing_references(file_path, ext)
                
                # Categorize by file type
                self.categorize_file_by_type(file_path, ext, size_mb)
    
    def find_duplicates(self):
        """Find files with duplicate names"""
        file_names = defaultdict(list)
        
        for root, dirs, files in os.walk(self.assets_path):
            for file in files:
                file_names[file].append(str(Path(root) / file))
        
        for name, paths in file_names.items():
            if len(paths) > 1:
                self.stats['duplicate_names'].append({
                    'name': name,
                    'locations': paths
                })
    
    def generate_report(self):
        """Generate comprehensive audit report"""
        print("\n" + "="*80)
        print("🎨 ASSET LIBRARY AUDIT REPORT")
        print("="*80)
        print(f"📅 Generated: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
        print(f"📁 Assets Directory: {self.assets_path}")
        print()
        
        # Basic Statistics
        print("📊 BASIC STATISTICS")
        print("-" * 40)
        print(f"📁 Total Directories: {self.stats['total_directories']}")
        print(f"📄 Total Files: {self.stats['total_files']}")
        print()
        
        # File Types
        print("📄 FILE TYPE BREAKDOWN")
        print("-" * 40)
        sorted_types = sorted(self.stats['file_types'].items(), key=lambda x: x[1], reverse=True)
        for ext, count in sorted_types:
            print(f"  {ext:>8}: {count:>4} files")
        print()
        
        # Asset Types
        print("🎨 ASSET TYPE BREAKDOWN")
        print("-" * 40)
        for asset_type, count in self.stats['asset_types'].items():
            if count > 0:
                print(f"  {asset_type:>12}: {count:>4} files")
        print()
        
        # Size Breakdown
        print("📏 SIZE BREAKDOWN")
        print("-" * 40)
        for size_category, count in self.stats['size_breakdown'].items():
            if count > 0:
                print(f"  {size_category:>12}: {count:>4} files")
        print()
        
        # Categories
        print("📂 DIRECTORY CATEGORIES")
        print("-" * 40)
        for category, count in self.stats['categories'].items():
            if count > 0:
                print(f"  {category:>12}: {count:>4} directories")
        print()
        
        # Largest Files
        print("📏 LARGEST FILES (Top 10)")
        print("-" * 40)
        largest_files = sorted(self.stats['largest_files'], key=lambda x: x['size_mb'], reverse=True)[:10]
        for file_info in largest_files:
            formatted_size = self.format_file_size(file_info['size_mb'])
            print(f"  {formatted_size:>10}: {file_info['path']}")
        print()
        
        # Directory Structure
        print("📁 DIRECTORY STRUCTURE")
        print("-" * 40)
        sorted_dirs = sorted(self.stats['directories'])
        for directory in sorted_dirs:
            print(f"  📁 {directory}")
        print()
        
        # Issues
        if self.stats['missing_files']:
            print("⚠️  MISSING FILES")
            print("-" * 40)
            for missing_file in self.stats['missing_files']:
                print(f"  ❌ {missing_file}")
            print()
        
        if self.stats['duplicate_names']:
            print("🔄 DUPLICATE FILE NAMES")
            print("-" * 40)
            for duplicate in self.stats['duplicate_names']:
                print(f"  📄 {duplicate['name']}")
                for location in duplicate['locations']:
                    print(f"      📁 {location}")
            print()
        
        # Summary
        total_size_mb = sum(file_info['size_mb'] for file_info in self.stats['largest_files'])
        print("📋 SUMMARY")
        print("-" * 40)
        print(f"📊 Total Library Size: {self.format_file_size(total_size_mb)}")
        print(f"📁 Directory Count: {self.stats['total_directories']}")
        print(f"📄 File Count: {self.stats['total_files']}")
        print(f"🎨 Asset Types: {sum(self.stats['asset_types'].values())}")
        print(f"⚠️  Issues Found: {len(self.stats['missing_files']) + len(self.stats['duplicate_names'])}")
        print()
        
        # Recommendations
        print("💡 RECOMMENDATIONS")
        print("-" * 40)
        if self.stats['missing_files']:
            print("  🔧 Fix missing referenced files")
        if self.stats['duplicate_names']:
            print("  🔧 Resolve duplicate file names")
        if self.stats['total_files'] > 1000:
            print("  📈 Consider implementing asset versioning")
        if total_size_mb > 1024:  # 1 GB
            print("  💾 Consider implementing asset compression")
        if self.stats['size_breakdown']['huge'] > 0 or self.stats['size_breakdown']['massive'] > 0:
            print("  🚀 Large files detected - consider cloud storage")
        if self.stats['asset_types']['audio'] > 0:
            print("  🎵 Audio files found - consider audio compression")
        if self.stats['asset_types']['video'] > 0:
            print("  🎬 Video files found - consider video compression")
        print("  ✅ Universal library audit complete!")
    
    def save_report(self, filename="asset_audit_report.json"):
        """Save audit report to JSON file"""
        report_data = {
            'timestamp': datetime.now().isoformat(),
            'assets_path': str(self.assets_path),
            'statistics': dict(self.stats),
            'file_types': dict(self.stats['file_types']),
            'categories': dict(self.stats['categories'])
        }
        
        with open(filename, 'w') as f:
            json.dump(report_data, f, indent=2)
        
        print(f"💾 Report saved to: {filename}")
    
    def run_audit(self):
        """Run complete asset audit"""
        print("🎨 Starting Asset Library Audit...")
        print(f"📁 Scanning: {self.assets_path}")
        print()
        
        # Scan assets directory
        self.scan_directory(self.assets_path)
        
        # Find duplicates
        print("🔍 Checking for duplicate files...")
        self.find_duplicates()
        
        # Generate report
        self.generate_report()
        
        # Save report
        self.save_report()
        
        print("✅ Asset audit complete!")

def main():
    """Main function"""
    auditor = AssetAuditor()
    auditor.run_audit()

if __name__ == "__main__":
    main() 

File: src/core/audit.cpp
/*
Author: KleaSCM
Email: KleaSCM@gmail.com
Name: audit.cpp
Description: Implementation of the AssetAuditor class. Provides modular, production-level asset library auditing for Blender Asset Manager. Implements scanning, statistics, reporting, and validation logic.
*/

#include "audit.hpp"
#include <iostream>
#include <fstream>
#include <algorithm>
#include <iomanip>
#include <chrono>
#include <sstream>

// JSON removed - console output only

// Constructor: Initializes auditor with project root and assets path
AssetAuditor::AssetAuditor(const std::filesystem::path& projectRoot_)
    : projectRoot(projectRoot_), assetsPath(projectRoot_ / "Assets") {}

// Run the complete audit: scan, find duplicates, report, save
void AssetAuditor::runAudit() {
    std::cout << "\U0001F3A8 Starting Asset Library Audit...\n";
    std::cout << "\U0001F4C1 Scanning: " << assetsPath << "\n\n";
    scanDirectory(assetsPath);
    std::cout << "\U0001F50D Checking for duplicate files...\n";
    findDuplicates();
    generateReport();
    saveReport("asset_audit_report.json");
    std::cout << "\u2705 Asset audit complete!\n";
}

// Recursively scan directory and collect statistics
void AssetAuditor::scanDirectory(const std::filesystem::path& directory) {
    if (!std::filesystem::exists(directory)) {
        std::cerr << "\u274C Directory not found: " << directory << std::endl;
        return;
    }
    for (const auto& entry : std::filesystem::recursive_directory_iterator(directory)) {
        if (entry.is_directory()) {
            stats.totalDirectories++;
            stats.directories.push_back(entry.path().lexically_relative(projectRoot).string());
            // Category detection (simplified for brevity)
            std::string dirLower = entry.path().filename().string();
            std::transform(dirLower.begin(), dirLower.end(), dirLower.begin(), ::tolower);
            if (dirLower.find("model") != std::string::npos || dirLower.find("building") != std::string::npos)
                stats.categories["Models"]++;
            else if (dirLower.find("texture") != std::string::npos || dirLower.find("material") != std::string::npos)
                stats.categories["Textures"]++;
            else if (dirLower.find("audio") != std::string::npos)
                stats.categories["Audio"]++;
            else if (dirLower.find("video") != std::string::npos)
                stats.categories["Video"]++;
            else if (dirLower.find("scene") != std::string::npos)
                stats.categories["Scenes"]++;
            else if (dirLower.find("script") != std::string::npos)
                stats.categories["Scripts"]++;
            else if (dirLower.find("doc") != std::string::npos)
                stats.categories["Documents"]++;
            else if (dirLower.find("archive") != std::string::npos)
                stats.categories["Archives"]++;
            else
                stats.categories[entry.path().filename().string()]++;
        } else if (entry.is_regular_file()) {
            stats.totalFiles++;
            std::string ext = entry.path().extension().string();
            std::transform(ext.begin(), ext.end(), ext.begin(), ::tolower);
            stats.fileTypes[ext]++;
            double sizeMB = getFileSizeMB(entry.path());
            stats.largestFiles.emplace_back(entry.path().lexically_relative(projectRoot).string(), sizeMB);
            checkForMissingReferences(entry.path(), ext);
            categorizeFileByType(entry.path(), ext, sizeMB);
        }
    }
}

// Find files with duplicate names
void AssetAuditor::findDuplicates() {
    std::map<std::string, std::vector<std::string>> fileNames;
    for (const auto& entry : std::filesystem::recursive_directory_iterator(assetsPath)) {
        if (entry.is_regular_file()) {
            std::string name = entry.path().filename().string();
            fileNames[name].push_back(entry.path().string());
        }
    }
    for (const auto& [name, paths] : fileNames) {
        if (paths.size() > 1) {
            stats.duplicateNames.emplace_back(name, paths);
        }
    }
}

// Check for missing referenced files (simplified universal logic)
void AssetAuditor::checkForMissingReferences(const std::filesystem::path& filePath, const std::string& ext) {
    static const std::map<std::string, std::vector<std::string>> referencePatterns = {
        {".mtl", {".obj"}},
        {".blend", {".png", ".jpg", ".jpeg", ".tga", ".tiff"}},
        {".fbx", {".png", ".jpg", ".jpeg", ".tga", ".tiff"}},
        {".gltf", {".png", ".jpg", ".jpeg", ".tga", ".tiff"}},
        {".glb", {".png", ".jpg", ".jpeg", ".tga", ".tiff"}},
    };
    auto it = referencePatterns.find(ext);
    if (it != referencePatterns.end()) {
        for (const auto& refExt : it->second) {
            auto refFile = filePath.parent_path() / (filePath.stem().string() + refExt);
            if (!std::filesystem::exists(refFile)) {
                bool found = false;
                for (const auto& subdir : std::filesystem::directory_iterator(filePath.parent_path())) {
                    if (subdir.is_directory()) {
                        auto potentialRef = subdir.path() / (filePath.stem().string() + refExt);
                        if (std::filesystem::exists(potentialRef)) {
                            found = true;
                            break;
                        }
                    }
                }
                if (!found) {
                    stats.missingFiles.push_back(refFile.lexically_relative(projectRoot).string());
                }
            }
        }
    }
}

// Categorize file by type and size
void AssetAuditor::categorizeFileByType(const std::filesystem::path& filePath, const std::string& ext, double sizeMB) {
    static const std::set<std::string> modelExts = {".blend", ".obj", ".fbx", ".dae", ".3ds", ".stl", ".ply", ".max", ".c4d", ".ma", ".mb", ".abc", ".usd", ".gltf", ".glb"};
    static const std::set<std::string> textureExts = {".png", ".jpg", ".jpeg", ".tga", ".tiff", ".bmp", ".exr", ".hdr", ".psd", ".ai", ".svg", ".webp", ".ktx", ".dds"};
    static const std::set<std::string> audioExts = {".mp3", ".wav", ".flac", ".aac", ".ogg", ".wma", ".m4a", ".aiff", ".au", ".mid", ".midi"};
    static const std::set<std::string> videoExts = {".mp4", ".avi", ".mov", ".wmv", ".flv", ".webm", ".mkv", ".m4v", ".3gp", ".ogv", ".ts", ".mts"};
    static const std::set<std::string> docExts = {".pdf", ".doc", ".docx", ".txt", ".rtf", ".md", ".html", ".xml", ".json", ".csv", ".xlsx", ".ppt", ".pptx"};
    static const std::set<std::string> archiveExts = {".zip", ".rar", ".7z", ".tar", ".gz", ".bz2", ".xz", ".dmg", ".iso"};
    static const std::set<std::string> scriptExts = {".py", ".js", ".php", ".rb", ".java", ".cpp", ".c", ".cs", ".sh", ".bat", ".ps1"};
    if (modelExts.count(ext)) stats.assetTypes["models"]++;
    else if (textureExts.count(ext)) stats.assetTypes["textures"]++;
    else if (audioExts.count(ext)) stats.assetTypes["audio"]++;
    else if (videoExts.count(ext)) stats.assetTypes["video"]++;
    else if (docExts.count(ext)) stats.assetTypes["documents"]++;
    else if (archiveExts.count(ext)) stats.assetTypes["archives"]++;
    else if (scriptExts.count(ext)) stats.assetTypes["scripts"]++;
    else stats.assetTypes["other"]++;
    if (sizeMB < 1) stats.sizeBreakdown["tiny"]++;
    else if (sizeMB < 10) stats.sizeBreakdown["small"]++;
    else if (sizeMB < 100) stats.sizeBreakdown["medium"]++;
    else if (sizeMB < 1024) stats.sizeBreakdown["large"]++;
    else if (sizeMB < 10240) stats.sizeBreakdown["huge"]++;
    else stats.sizeBreakdown["massive"]++;
}

// Get file size in MB
double AssetAuditor::getFileSizeMB(const std::filesystem::path& filePath) const {
    try {
        return static_cast<double>(std::filesystem::file_size(filePath)) / (1024.0 * 1024.0);
    } catch (...) {
        return 0.0;
    }
}

// Format file size for display
std::string AssetAuditor::formatFileSize(double sizeMB) const {
    std::ostringstream oss;
    if (sizeMB >= 1024 * 1024)
        oss << std::fixed << std::setprecision(2) << (sizeMB / (1024 * 1024)) << " TB";
    else if (sizeMB >= 1024)
        oss << std::fixed << std::setprecision(2) << (sizeMB / 1024) << " GB";
    else if (sizeMB >= 1)
        oss << std::fixed << std::setprecision(2) << sizeMB << " MB";
    else
        oss << std::fixed << std::setprecision(0) << (sizeMB * 1024) << " KB";
    return oss.str();
}

// Generate a human-readable report to stdout
void AssetAuditor::generateReport() const {
    std::cout << "\n" << std::string(80, '=') << "\n";
    std::cout << "\U0001F3A8 ASSET LIBRARY AUDIT REPORT\n";
    std::cout << std::string(80, '=') << "\n";
    std::cout << "\U0001F4C5 Generated: " << getTimestamp() << "\n";
    std::cout << "\U0001F4C1 Assets Directory: " << assetsPath << "\n\n";
    std::cout << "\U0001F4CA BASIC STATISTICS\n" << std::string(40, '-') << "\n";
    std::cout << "\U0001F4C1 Total Directories: " << stats.totalDirectories << "\n";
    std::cout << "\U0001F4C4 Total Files: " << stats.totalFiles << "\n\n";
    std::cout << "\U0001F4C4 FILE TYPE BREAKDOWN\n" << std::string(40, '-') << "\n";
    for (const auto& [ext, count] : stats.fileTypes)
        std::cout << "  " << std::setw(8) << ext << ": " << std::setw(4) << count << " files\n";
    std::cout << "\n\U0001F3A8 ASSET TYPE BREAKDOWN\n" << std::string(40, '-') << "\n";
    for (const auto& [type, count] : stats.assetTypes)
        if (count > 0) std::cout << "  " << std::setw(12) << type << ": " << std::setw(4) << count << " files\n";
    std::cout << "\n\U0001F4CF SIZE BREAKDOWN\n" << std::string(40, '-') << "\n";
    for (const auto& [cat, count] : stats.sizeBreakdown)
        if (count > 0) std::cout << "  " << std::setw(12) << cat << ": " << std::setw(4) << count << " files\n";
    std::cout << "\n\U0001F4C2 DIRECTORY CATEGORIES\n" << std::string(40, '-') << "\n";
    for (const auto& [cat, count] : stats.categories)
        if (count > 0) std::cout << "  " << std::setw(12) << cat << ": " << std::setw(4) << count << " directories\n";
    std::cout << "\n\U0001F4CF LARGEST FILES (Top 10)\n" << std::string(40, '-') << "\n";
    auto largest = stats.largestFiles;
    std::sort(largest.begin(), largest.end(), [](const auto& a, const auto& b) { return a.second > b.second; });
    for (size_t i = 0; i < std::min<size_t>(10, largest.size()); ++i)
        std::cout << "  " << std::setw(10) << formatFileSize(largest[i].second) << ": " << largest[i].first << "\n";
    std::cout << "\n\U0001F4C1 DIRECTORY STRUCTURE\n" << std::string(40, '-') << "\n";
    auto dirs = stats.directories;
    std::sort(dirs.begin(), dirs.end());
    for (const auto& dir : dirs)
        std::cout << "  \U0001F4C1 " << dir << "\n";
    if (!stats.missingFiles.empty()) {
        std::cout << "\n\u26A0\uFE0F  MISSING FILES\n" << std::string(40, '-') << "\n";
        for (const auto& f : stats.missingFiles)
            std::cout << "  \u274C " << f << "\n";
    }
    if (!stats.duplicateNames.empty()) {
        std::cout << "\n\U0001F501 DUPLICATE FILE NAMES\n" << std::string(40, '-') << "\n";
        for (const auto& [name, locations] : stats.duplicateNames) {
            std::cout << "  \U0001F4C4 " << name << "\n";
            for (const auto& loc : locations)
                std::cout << "      \U0001F4C1 " << loc << "\n";
        }
    }
    double totalSizeMB = 0.0;
    for (const auto& f : stats.largestFiles) totalSizeMB += f.second;
    std::cout << "\n\U0001F4CB SUMMARY\n" << std::string(40, '-') << "\n";
    std::cout << "\U0001F4CA Total Library Size: " << formatFileSize(totalSizeMB) << "\n";
    std::cout << "\U0001F4C1 Directory Count: " << stats.totalDirectories << "\n";
    std::cout << "\U0001F4C4 File Count: " << stats.totalFiles << "\n";
    std::cout << "\U0001F3A8 Asset Types: " << stats.assetTypes.size() << "\n";
    std::cout << "\u26A0\uFE0F  Issues Found: " << (stats.missingFiles.size() + stats.duplicateNames.size()) << "\n\n";
    std::cout << "\U0001F4A1 RECOMMENDATIONS\n" << std::string(40, '-') << "\n";
    if (!stats.missingFiles.empty()) std::cout << "  \U0001F527 Fix missing referenced files\n";
    if (!stats.duplicateNames.empty()) std::cout << "  \U0001F527 Resolve duplicate file names\n";
    if (stats.totalFiles > 1000) std::cout << "  \U0001F4C8 Consider implementing asset versioning\n";
    if (totalSizeMB > 1024) std::cout << "  \U0001F4BE Consider implementing asset compression\n";
    if (stats.sizeBreakdown.count("huge") && stats.sizeBreakdown.at("huge") > 0)
        std::cout << "  \U0001F680 Large files detected - consider cloud storage\n";
    if (stats.assetTypes.count("audio") && stats.assetTypes.at("audio") > 0)
        std::cout << "  \U0001F3B5 Audio files found - consider audio compression\n";
    if (stats.assetTypes.count("video") && stats.assetTypes.at("video") > 0)
        std::cout << "  \U0001F3AC Video files found - consider video compression\n";
    std::cout << "  \u2705 Universal library audit complete!\n";
}

// Save audit report as simple text file
void AssetAuditor::saveReport(const std::string& filename) const {
    std::ofstream out(filename);
    out << "Asset Audit Report - " << getTimestamp() << std::endl;
    out << "Assets Path: " << assetsPath << std::endl;
    out << "Total Files: " << stats.totalFiles << std::endl;
    out << "Total Directories: " << stats.totalDirectories << std::endl;
    out << "Issues Found: " << (stats.missingFiles.size() + stats.duplicateNames.size()) << std::endl;
    std::cout << "\U0001F4BE Report saved to: " << filename << std::endl;
}

// Utility: Get current timestamp as string
std::string AssetAuditor::getTimestamp() const {
    auto now = std::chrono::system_clock::now();
    std::time_t t = std::chrono::system_clock::to_time_t(now);
    char buf[32];
    std::strftime(buf, sizeof(buf), "%Y-%m-%d %H:%M:%S", std::localtime(&t));
    return std::string(buf);
} 